server/controller/src/main/java/ai/starwhale/mlops/datastore/ColumnType.java

```diff
@@ -89,102 +101,168 @@ public static int compare(ColumnType type1, Object value1, ColumnType type2, Obj
+                                    .compareTo(BigDecimal.valueOf(((Number) value2).doubleValue()));
                         }
-                        return Double.compare(((Number) value1).doubleValue(), ((Number) value2).doubleValue());
-                    } else if (category1.equals(ColumnTypeScalar.FLOAT32.getCategory())
-                            && category2.equals(ColumnTypeScalar.INT32.getCategory())) {
-                        return -ColumnType.compare(type2, value2, type1, value1);
                     }
+                    return Double.compare(((Number) value1).doubleValue(), ((Number) value2).doubleValue());
+                } else if (category1.equals(ColumnTypeScalar.FLOAT32.getCategory())
+                        && category2.equals(ColumnTypeScalar.INT32.getCategory())) {
+                    return -ColumnType.compare(type2, value2, type1, value1);
                 }
             }
-        } else if (type1 instanceof ColumnTypeList && type2 instanceof ColumnTypeList) {
-            var iter1 = ((List<?>) value1).iterator();
-            var iter2 = ((List<?>) value2).iterator();
-            for (; ; ) {
-                if (iter1.hasNext() && iter2.hasNext()) {
-                    var result = ColumnType.compare(((ColumnTypeList) type1).getElementType(), iter1.next(),
-                            ((ColumnTypeList) type2).getElementType(), iter2.next());
-                    if (result != 0) {
-                        return result;
-                    }
-                } else if (iter1.hasNext()) {
-                    return 1;
-                } else if (iter2.hasNext()) {
-                    return -1;
+        }
+        throw new IllegalArgumentException("can not compare " + type1 + " with " + type2);
+    }
+
+    private static int compareList(ColumnTypeList type1, Object value1, ColumnTypeList type2, Object value2) {
+        var iter1 = ((List<?>) value1).iterator();
+        var iter2 = ((List<?>) value2).iterator();
+        for (; ; ) {
+            if (iter1.hasNext() && iter2.hasNext()) {
+                var result = ColumnType.compare(type1.getElementType(),
+                        iter1.next(),
+                        type2.getElementType(),
+                        iter2.next());
+                if (result != 0) {
+                    return result;
+                }
+            } else if (iter1.hasNext()) {
+                return 1;
+            } else if (iter2.hasNext()) {
+                return -1;
+            } else {
+                return 0;
+            }
+        }
+    }
+
+    private static int compareMap(ColumnTypeMap type1, Object value1, ColumnTypeMap type2, Object value2) {
+        var map1 = (Map<?, ?>) value1;
+        var map2 = (Map<?, ?>) value2;
+        if (map1.isEmpty() && map2.isEmpty()) {
+            return 0;
+        }
+        if (map1.isEmpty()) {
+            return -1;
+        }
+        if (map2.isEmpty()) {
+            return 1;
+        }
```

Comment:
This block is just a fast path? can we remove it?

Reply:
Changed to another algorithm implementation.